### ISSUE TITLE
How to figure out why a reader is not found

### ISSUE THREAD

**Issue by travisby** [GoogleChromeLabs/chromeos_smart_card_connector#83]
Hello,

tl;dr: My card reader is detected in ChromeOS but not Chrome on Fedora or Windows.  I'd love some help on figuring out where to add debugging information, since the bridge between C and javascript goes a little over my head.

Context
---------

I own a Gemalto PC Pinpad (reads: Ezio shield -- [lsusb output](https://gist.github.com/travisby/1d58a870360fcf48b3729a2ef9be0eb7)) that I use with an OpenPGP Card (v2.x).

Under ChromiumOS (CloudReady) and I believe also ChromeOS v75 via a pixel slate the Smart Card Connector App is able to detect and connnect to my device (even sending some sample pcsc commands, and ccid commands).

Unfortunately it's not quite detected under Fedora or Windows with Chrome/Chromium using the same extension.  The reader works with both pcscd and gpg/scdaemon in this environment normally.

What I've tried/found
-------------------

Following [Troubleshooting Apps under desktop OSes](https://github.com/GoogleChromeLabs/chromeos_smart_card_connector#troubleshooting-apps-under-desktop-oses)  I killed pcscd and checked the permissions on the device and setup some udev rules:

```bash
[travis@travis ~]$ ls /dev/bus/usb/001/016 -l
crw-rw-rw-. 1 root root 189, 15 Jul 30 08:19 /dev/bus/usb/001/016
[travis@travis ~]$ ls /dev/bus/usb/001/016 -lZ
crw-rw-rw-. 1 root root system_u:object_r:usb_device_t:s0 189, 15 Jul 30 08:19 /dev/bus/usb/001/016
[travis@travis ~]$ ps aux | grep pcscd
travis   24930  0.0  0.0   5992   828 pts/1    S+   08:24   0:00 grep --color=auto pcscd
[travis@travis ~]$
```

still with no luck.

`background.js` shows the reader is found by libusb:
```
 [  0.061s] [GoogleSmartCard.Libusb.ChromeUsbBackend] 1 USB device(s) available: [{"device": 0x00, "manufacturerName": "Gemalto", "productId": 0x000034C2, "productName": "Ezio Shield", "serialNumber": "R145023804255", "vendorId": 0x000008E6, "version": 0x00000102}]
```

But later `window.js` will remind us that there's no hope here:

```
[  0.012s] [GoogleSmartCard.ConnectorApp.MainWindow] Displaying 0 card reader(s): []
```

I've rebuilt the app with `export CONFIG=Debug` and I don't think I was seeing any more log messages, so I also ran `find smart_card_connector_app -exec sed -i 's/logger.fine/logger.info/g' {} \;` for good measure.  There's definitely more logging now but pretty much only helpful for Known Apps.

Note that neither ChromeOS or GNU/Linux have the device appear in the "Add Device" window, because that [searches exclusively for 0x0B interface-class smart card readers](https://github.com/GoogleChromeLabs/chromeos_smart_card_connector/blob/42866b5aa0c7928ff431ad1da6c7c20bd0937e78/smart_card_connector_app/src/window-devices-displaying.js#L126), which approximately [5% of readers are not](https://ludovicrousseau.blogspot.com/2013/06/ccid-descriptor-statistics_7.html), but the ReaderTracker list seems to be getting populated from elsewhere, and seems sufficient for the usecase. (I did `window.readerTracker = readerTracker` somewhere and tried playing around with debugging that way to no avail.  Both the "successful" and "unsuccessful" readers are empty)

I'm not quite sure how to debug further, I'd love some suggestions on where I could add logging to help narrow down why my reader works in one environment but not the other.

**Comment by emaxx-google** [comment]
Hello,
Try obtaining the logs via the "Export logs" button.

**Comment by travisby** [comment]
Thanks!  I updated the gist with the [export logs](https://gist.github.com/travisby/1d58a870360fcf48b3729a2ef9be0eb7) button (and then copy&paste).  I bet that would have been better than all the `sed` magic :).  Unfortunately I'm not seeing anything telling there.


```
 [  0.029s] [<background page>.GoogleSmartCard.Libusb.ChromeUsbBackend] [INFO] 1 USB device(s) available: [{"device": 0x00, "manufacturerName": "Gemalto", "productId": 0x000034C2, "productName": "Ezio Shield", "serialNumber": "R145023804255", "vendorId": 0x000008E6, "version": 0x00000102}]
....
 [  8.196s] [</window.html>.GoogleSmartCard.ConnectorApp.MainWindow] [INFO] Application list updated, refreshing the view. New list of id's: []
```
seem to still show the discrepency between usb and pcsc without much in the way of errors in between.  I'll try to find where pcsc gets its list of usb devices from (is it before or after ccid?)

**Comment by emaxx-google** [comment]
Looks like the NaCl module simply doesn't work: the logs don't contain the message "Successfully loaded NaCl module".

Try going to chrome://components/ and checking if "pnacl" is installed.

**Comment by travisby** [comment]
That must be it!  I don't see anyythign like nacl in `chrome://components`.  Thank you!  I'll figure out how to get that enabled/installed and report back

**Comment by travisby** [comment]
Interestingly enough, switching to google-chrome and installing the smart-card-connector-app from the store ended up causing a restart loop for the app.  Installing my local development version yielded "pnacl not installed" error.  It was listed in `chrome://components` but was at version 0.*.  Manually hitting update led to the app working and detecting my reader!

I'm guessing why it worked on my slate was official google-chrome, and why it also worked now on fedora after switching to chrome from chromium. 

 I'm still confused why CloudReady which would have been using Chromium but worked without pnacl listed in the components section, and why the error condition was much more graceful in google-chrome ("pnacl not installed" in console.error) vs the subtle failure we've noticed here in chromium.  Maybe Chrom*OS is using a different method than pnacl since the browser is so tied to the OS.

Separately I'm guessing (and still investigating) that the deprecation of pnacl outside of ChromeOS is why it didn't work on fedora chromium (Maybe they removed it in upstream but google-chrome is still holding on to the component until the full deprecation date, but chromium removed it early?  I'm not sure, going to keep digging for my own amusement).

But none of these are smart-card-connector-app problems (except maybe the extension crash-looping if pnacl isn't installed, but I think that gets updated in the background so that's probably an uncommon issue).  Thank you so much for your help :).

**Comment by emaxx-google** [comment]
Glad to hear that the problem is resolved.

I don't know much about non-official Chromium browsers and why they don't have NaCl - as in theory Native Client is an open-source technology - but if that's so, then probably the reason why the error was silent was that the "application/x-pnacl" <embed>'s MIME type was unknown and therefore the whole embed was ignored.

Probably I could add some heuristic that if the <embed> doesn't send any event in, say, 1 minute, then emit a corresponding error message. But it'd be a low-priority feature request, since we consider the Chrome OS as the only target platform for the Smart Card Connector app (for multiple reasons, for example, that desktop platforms already have existing smart card infrastructure, which Chrome already knows how to talk to).

In your other environment, where the "pnacl not installed" error was emitted, the crash-restart loop is also "works as intended", since the app is programmed to try to recover from fatal errors by restarting itself when running in the Release mode. Although, admittedly, we could enhance this logic - this is tracked by feature request #36.

P.S. And, yes, the NaCl technology is deprecated in favor of the more portable WebAssembly one - we're currently planning the work on the migration of the Smart Card Connector app.
